src/Symfony/Component/DependencyInjection/ContainerBuilder.php

```diff
@@ -541,14 +543,25 @@ public function compile()
         $compiler->compile($this);
         $this->compiled = true;
 
+        $renameDefinitions = array();
         foreach ($this->definitions as $id => $definition) {
             if (!$definition->isPublic()) {
-                $this->privates[$id] = true;
+                if (null === $privateOriginId = $definition->getPrivateOriginId()) {
```

Comment:
This should be kept within a compiler pass, as depends on the existence of another Compiler pass (which could be disabled).